discovery/manager.go

```diff
@@ -153,41 +146,38 @@ func (m *Manager) cancelDiscoverers() {
 	for _, c := range m.discoverCancel {
 		c()
 	}
-	m.targets = make(map[poolKey][]*targetgroup.Group)
+	m.targets = make(map[poolKey]map[string]*targetgroup.Group)
 	m.discoverCancel = nil
 }
 
-func (m *Manager) addGroup(poolKey poolKey, tg []*targetgroup.Group) {
+func (m *Manager) updateGroup(poolKey poolKey, tgs []*targetgroup.Group) {
 	done := make(chan struct{})
 
 	m.actionCh <- func(ctx context.Context) {
-		if tg != nil {
-			m.targets[poolKey] = tg
+		if tgs != nil {
+			for _, tg := range tgs {
+				if tg != nil { // Some Discoverers send nil targetgroup so need to check for it to avoid panics.
```

Comment:
nit but i found this more readable:

```
if tg == nil {
    continue
}
...
```

Reply:
And FWIW I've spotted the places in the K8S SD where it sends nil target groups (see https://github.com/prometheus/prometheus/pull/3660) but it doesn't mean that all SD providers are safe either.

Reply:
thanks